File: scgraph/cpp/src/graph_reducer.cpp
#include <queue>
#include <algorithm>
#include <tuple>
#include "graph_reducer.hpp"
#include "contraction_hierarchies.hpp"
#include "transit_node_routing.hpp"

void GraphReducer::reset_cache() {
    GraphUtils::reset_cache();
    has_reduced_graph = false;
    is_reduced.clear();
    reduced_node_chain_ids.clear();
    reduced_graph.clear();
    reduced_graph_connections.clear();
    reduced_inverse_graph.clear();
    reduced_inverse_graph_connections.clear();
}

void GraphReducer::reduce() {
    this->reset_cache();
    this->ensure_inverse_graph();

    size_t n = graph.size();
    is_reduced.assign(n, false);

    for (size_t u = 0; u < n; ++u) {
        // Count outflows excluding self-loop
        size_t outflows_len = 0;
        for (const auto& edge : graph[u]) {
            if (edge.first != (int)u) {
                outflows_len++;
            }
        }

        size_t inflows_len = 0;
        for (const auto& edge : inverse_graph[u]) {
            if (edge.first != (int)u) {
                inflows_len++;
            }
        }

        if (outflows_len == 1) {
            if (inflows_len >= 1) {
                is_reduced[u] = true;
            }
        } else if (outflows_len == 2) {
            if (inflows_len > 0) {
                // Get set of outflows and inflows
                std::set<int> outflows;
                for (const auto& edge : graph[u]) {
                    if (edge.first != (int)u) {
                        outflows.insert(edge.first);
                    }
                }
                bool subset = true;
                for (const auto& edge : inverse_graph[u]) {
                    if (edge.first != (int)u) {
                        if (outflows.find(edge.first) == outflows.end()) {
                            subset = false;
                            break;
                        }
                    }
                }
                if (subset) {
                    is_reduced[u] = true;
                }
            }
        }
    }

    // Assign chain IDs to connected components of reduced nodes
    reduced_node_chain_ids.assign(n, -1);
    int current_chain_id = 0;
    for (size_t u = 0; u < n; ++u) {
        if (is_reduced[u] && reduced_node_chain_ids[u] == -1) {
            std::vector<int> q = {(int)u};
            reduced_node_chain_ids[u] = current_chain_id;
            while (!q.empty()) {
                int curr = q.back();
                q.pop_back();
                for (const auto& edge : graph[curr]) {
                    int v = edge.first;
                    if (v != curr && v >= 0 && v < (int)n && is_reduced[v] && reduced_node_chain_ids[v] == -1) {
                        reduced_node_chain_ids[v] = current_chain_id;
                        q.push_back(v);
                    }
                }
                for (const auto& edge : inverse_graph[curr]) {
                    int v = edge.first;
                    if (v != curr && v >= 0 && v < (int)n && is_reduced[v] && reduced_node_chain_ids[v] == -1) {
                        reduced_node_chain_ids[v] = current_chain_id;
                        q.push_back(v);
                    }
                }
            }
            current_chain_id++;
        }
    }

    // 3. Build reduced graph and connections (outbound from all nodes)
    reduced_graph.assign(n, std::vector<std::pair<int, double>>());
    reduced_graph_connections.assign(n, std::unordered_map<int, std::vector<int>>());

    for (size_t A = 0; A < n; ++A) {
        std::unordered_map<int, double> best_dist;
        using PQElement = std::tuple<double, int, std::vector<int>>;
        std::priority_queue<PQElement, std::vector<PQElement>, std::greater<PQElement>> open_leaves;

        open_leaves.push({0.0, (int)A, {}});

        while (!open_leaves.empty()) {
            auto [dist, u, path] = open_leaves.top();
            open_leaves.pop();

            if (best_dist.find(u) != best_dist.end() && best_dist[u] <= dist) {
                continue;
            }
            best_dist[u] = dist;

            if (u != (int)A && !is_reduced[u]) {
                // Boundary non-reduced node. Record connection.
                reduced_graph[A].push_back({u, dist});
                if (!path.empty()) {
                    reduced_graph_connections[A][u] = path;
                }
                continue;
            }

            for (const auto& edge : graph[u]) {
                int v = edge.first;
                double w = edge.second;
                double new_dist = dist + w;
                if (best_dist.find(v) == best_dist.end() || new_dist < best_dist[v]) {
                    std::vector<int> new_path = (u == (int)A) ? std::vector<int>{} : path;
                    if (u != (int)A) {
                        new_path.push_back(u);
                    }
                    open_leaves.push({new_dist, v, new_path});
                }
            }
        }
    }

    // 4. Build reduced inverse graph and connections (inbound into all nodes)
    reduced_inverse_graph.assign(n, std::vector<std::pair<int, double>>());
    reduced_inverse_graph_connections.assign(n, std::unordered_map<int, std::vector<int>>());

    for (size_t B = 0; B < n; ++B) {
        std::unordered_map<int, double> best_dist;
        using PQElement = std::tuple<double, int, std::vector<int>>;
        std::priority_queue<PQElement, std::vector<PQElement>, std::greater<PQElement>> open_leaves;

        open_leaves.push({0.0, (int)B, {}});

        while (!open_leaves.empty()) {
            auto [dist, u, path] = open_leaves.top();
            open_leaves.pop();

            if (best_dist.find(u) != best_dist.end() && best_dist[u] <= dist) {
                continue;
            }
            best_dist[u] = dist;

            if (u != (int)B && !is_reduced[u]) {
                // Boundary non-reduced node reaching B.
                reduced_inverse_graph[B].push_back({u, dist});
                std::vector<int> fwd_path = path;
                std::reverse(fwd_path.begin(), fwd_path.end());
                if (!fwd_path.empty()) {
                    reduced_inverse_graph_connections[B][u] = fwd_path;
                }
                continue;
            }

            for (const auto& edge : inverse_graph[u]) {
                int v = edge.first;
                double w = edge.second;
                double new_dist = dist + w;
                if (best_dist.find(v) == best_dist.end() || new_dist < best_dist[v]) {
                    std::vector<int> new_path = (u == (int)B) ? std::vector<int>{} : path;
                    if (u != (int)B) {
                        new_path.push_back(u);
                    }
                    open_leaves.push({new_dist, v, new_path});
                }
            }
        }
    }

    has_reduced_graph = true;
}

std::vector<int> GraphReducer::expand_path(const std::vector<int>& path) const {
    if (!has_reduced_graph) {
        return path;
    }
    std::vector<int> new_path;
    for (size_t i = 0; i < path.size() - 1; ++i) {
        int u = path[i];
        int v = path[i + 1];
        new_path.push_back(u);
        bool expanded = false;
        if (u >= 0 && u < (int)reduced_graph_connections.size()) {
            const auto& conns = reduced_graph_connections[u];
            auto it = conns.find(v);
            if (it != conns.end()) {
                new_path.insert(new_path.end(), it->second.begin(), it->second.end());
                expanded = true;
            }
        }
        if (!expanded && v >= 0 && v < (int)reduced_inverse_graph_connections.size()) {
            const auto& inv_conns = reduced_inverse_graph_connections[v];
            auto it = inv_conns.find(u);
            if (it != inv_conns.end()) {
                new_path.insert(new_path.end(), it->second.begin(), it->second.end());
            }
        }
    }
    if (!path.empty()) {
        new_path.push_back(path.back());
    }
    return new_path;
}

std::vector<std::unordered_map<int, double>> GraphReducer::get_reduced_graph() const {
    std::vector<std::unordered_map<int, double>> result;
    result.reserve(reduced_graph.size());
    for (size_t i = 0; i < reduced_graph.size(); ++i) {
        std::unordered_map<int, double> adj;
        for (const auto& [v, w] : reduced_graph[i]) {
            adj[v] = w;
        }
        result.push_back(std::move(adj));
    }
    return result;
}

std::vector<std::unordered_map<int, double>> GraphReducer::get_reduced_inverse_graph() const {
    std::vector<std::unordered_map<int, double>> result;
    result.reserve(reduced_inverse_graph.size());
    for (size_t i = 0; i < reduced_inverse_graph.size(); ++i) {
        std::unordered_map<int, double> adj;
        for (const auto& [v, w] : reduced_inverse_graph[i]) {
            adj[v] = w;
        }
        result.push_back(std::move(adj));
    }
    return result;
}

bool GraphReducer::is_same_chain(const std::variant<int, std::set<int>>& origin_id, std::optional<int> destination_id) const {
    if (!destination_id.has_value() || reduced_node_chain_ids.empty()) {
        return false;
    }
    int dest = destination_id.value();
    if (dest < 0 || dest >= (int)reduced_node_chain_ids.size()) {
        return false;
    }
    int dest_chain = reduced_node_chain_ids[dest];
    if (dest_chain == -1) {
        return false;
    }
    if (std::holds_alternative<int>(origin_id)) {
        int orig = std::get<int>(origin_id);
        if (orig >= 0 && orig < (int)reduced_node_chain_ids.size()) {
            return reduced_node_chain_ids[orig] == dest_chain;
        }
        return false;
    } else {
        const auto& origins = std::get<std::set<int>>(origin_id);
        for (int orig : origins) {
            if (orig >= 0 && orig < (int)reduced_node_chain_ids.size() && reduced_node_chain_ids[orig] == dest_chain) {
                return true;
            }
        }
        return false;
    }
}

std::function<double(CHGraph*, int)> GraphReducer::wrap_heuristic(std::function<double(CHGraph*, int)> heuristic_fn) const {
    if (!has_reduced_graph) {
        return heuristic_fn;
    }
    if (!heuristic_fn) {
        return [this](CHGraph* ch, int node_id) {
            return (is_reduced[node_id] ? 0.0 : 1000000.0) + ch->default_heuristic(node_id);
        };
    }
    return [this, heuristic_fn](CHGraph* ch, int node_id) {
        return (is_reduced[node_id] ? 0.0 : 1000000.0) + heuristic_fn(ch, node_id);
    };
}

std::shared_ptr<CHGraph> GraphReducer::create_ch_graph(std::function<double(CHGraph*, int)> heuristic_fn, int settled_limit) const {
    if (has_reduced_graph) {
        return std::make_shared<CHGraph>(get_reduced_graph(), settled_limit, wrap_heuristic(heuristic_fn), get_reduced_inverse_graph());
    }
    return std::make_shared<CHGraph>(get_graph(), settled_limit, heuristic_fn);
}

std::shared_ptr<TNRGraph> GraphReducer::create_tnr_graph(int num_transit_nodes, std::function<double(CHGraph*, int)> heuristic_fn, int settled_limit) const {
    if (has_reduced_graph) {
        return std::make_shared<TNRGraph>(get_reduced_graph(), settled_limit, num_transit_nodes, wrap_heuristic(heuristic_fn), get_reduced_inverse_graph());
    }
    return std::make_shared<TNRGraph>(get_graph(), settled_limit, num_transit_nodes, heuristic_fn);
}

File: scgraph/graph_reducer.py
from heapq import heappop, heappush
from functools import wraps


def use_reduced(func):
    @wraps(func)
    def wrapper(self, *args, **kwargs):
        if getattr(self, "reduced_graph", None) is None:
            return func(self, *args, **kwargs)
        return self.__run_with_reduced__(func, *args, **kwargs)

    return wrapper


class GraphReducer:
    def reduce(self) -> None:
        """
        Function:

        - Reduce the graph by bypassing pass-through nodes and summing intermediate weights.
        - Sets self.reduced_graph, self.reduced_graph_connections, and self.is_reduced.

        Required Arguments:

        - None

        Returns:

        - None
        """
        self.reset_cache()
        self.__ensure_inverse_graph__()

        # 1. Identify pass-through nodes
        self.is_reduced = [False] * len(self.__graph__)
        graph = self.__graph__
        inverse_graph = self.__inverse_graph__
        is_reduced = self.is_reduced

        for u in range(len(graph)):
            out_dict = graph[u]
            in_dict = inverse_graph[u]

            out_len = len(out_dict) - (1 if u in out_dict else 0)
            if out_len == 1:
                in_len = len(in_dict) - (1 if u in in_dict else 0)
                if in_len >= 1:
                    is_reduced[u] = True
            elif out_len == 2:
                in_len = len(in_dict) - (1 if u in in_dict else 0)
                if in_len > 0:
                    outflows = {v for v in out_dict if v != u}
                    inflows = {v for v in in_dict if v != u}
                    if inflows.issubset(outflows):
                        is_reduced[u] = True

        # 2. Assign chain IDs to connected components of reduced nodes
        self.reduced_node_chain_ids = [None] * len(graph)
        reduced_node_chain_ids = self.reduced_node_chain_ids
        current_chain_id = 0
        for u in range(len(graph)):
            if is_reduced[u] and reduced_node_chain_ids[u] is None:
                queue = [u]
                reduced_node_chain_ids[u] = current_chain_id
                while queue:
                    curr = queue.pop()
                    for v in graph[curr]:
                        if (
                            v != curr
                            and is_reduced[v]
                            and reduced_node_chain_ids[v] is None
                        ):
                            reduced_node_chain_ids[v] = current_chain_id
                            queue.append(v)
                    for v in inverse_graph[curr]:
                        if (
                            v != curr
                            and is_reduced[v]
                            and reduced_node_chain_ids[v] is None
                        ):
                            reduced_node_chain_ids[v] = current_chain_id
                            queue.append(v)
                current_chain_id += 1

        # 3. Build reduced graph and connections (outbound from all nodes)
        self.reduced_graph = [{} for _ in range(len(graph))]
        self.reduced_graph_connections = [None] * len(graph)
        reduced_graph = self.reduced_graph
        reduced_graph_connections = self.reduced_graph_connections

        for A in range(len(graph)):
            best_dist = {}
            open_leaves = [(0, A, [])]

            while open_leaves:
                dist, u, path = heappop(open_leaves)
                if u in best_dist and best_dist[u] <= dist:
                    continue
                best_dist[u] = dist

                if u != A and not is_reduced[u]:
                    # Boundary non-reduced node. Record connection.
                    reduced_graph[A][u] = dist
                    if path:
                        if reduced_graph_connections[A] is None:
                            reduced_graph_connections[A] = {}
                        reduced_graph_connections[A][u] = path
                    continue

                for v, w in graph[u].items():
                    new_dist = dist + w
                    if v not in best_dist or new_dist < best_dist[v]:
                        new_path = path if u == A else path + [u]
                        heappush(open_leaves, (new_dist, v, new_path))

        # 4. Build reduced inverse graph and connections (inbound into all nodes)
        self.reduced_inverse_graph = [{} for _ in range(len(graph))]
        self.reduced_inverse_graph_connections = [None] * len(graph)
        reduced_inverse_graph = self.reduced_inverse_graph
        reduced_inverse_graph_connections = (
            self.reduced_inverse_graph_connections
        )

        for B in range(len(graph)):
            best_dist = {}
            open_leaves = [(0, B, [])]

            while open_leaves:
                dist, u, path = heappop(open_leaves)
                if u in best_dist and best_dist[u] <= dist:
                    continue
                best_dist[u] = dist

                if u != B and not is_reduced[u]:
                    # Boundary non-reduced node reaching B
                    reduced_inverse_graph[B][u] = dist
                    fwd_path = list(reversed(path))
                    if fwd_path:
                        if reduced_inverse_graph_connections[B] is None:
                            reduced_inverse_graph_connections[B] = {}
                        reduced_inverse_graph_connections[B][u] = fwd_path
                    continue

                for v, w in inverse_graph[u].items():
                    new_dist = dist + w
                    if v not in best_dist or new_dist < best_dist[v]:
                        new_path = path if u == B else path + [u]
                        heappush(open_leaves, (new_dist, v, new_path))

    def wrap_heuristic(self, heuristic_fn=None):
        """
        Function:

        - Wrap a CH/TNR heuristic function to contract reduced nodes before non-reduced nodes.
        """
        if not getattr(self, "is_reduced", None):
            return heuristic_fn
        if heuristic_fn is None:
            is_reduced = self.is_reduced
            return lambda ch, n: (
                0 if is_reduced[n] else 1000000
            ) + ch.default_heuristic(n)
        is_reduced = self.is_reduced
        return lambda ch, n: (0 if is_reduced[n] else 1000000) + heuristic_fn(
            ch, n
        )

    def __run_with_reduced__(self, func, *args, **kwargs):
        """
        Function:

        - Run a graph routing function on the reduced graph.
        - Same-chain queries route directly on the unreduced graph.
        - Bidirectional algorithms route on reduced_graph and reduced_inverse_graph.
        - One-sided algorithms with reduced destination solve via boundary entry nodes from reduced_inverse_graph.

        Required Arguments:

        - `func`
            - Type: function
            - What: The routing algorithm function to execute

        Returns:

        - The return value of the wrapped routing algorithm (typically a dict with 'path' and 'length')
        """
        origin_id = kwargs.get("origin_id")
        if origin_id is None and len(args) > 0:
            origin_id = args[0]

        destination_id = kwargs.get("destination_id")
        if destination_id is None and len(args) > 1:
            destination_id = args[1]

        # 1. Same-chain check: solve on unreduced graph with dijkstra
        if destination_id is not None and self.is_same_chain(
            origin_id, destination_id
        ):
            length_only = kwargs.get("length_only", False)
            res = self.__dijkstra_on_graph__(
                self.__graph__, origin_id, destination_id
            )
            if length_only:
                res.pop("path", None)
            return res

        is_reduced = self.is_reduced

        # 2. Bidirectional algorithms, Contraction Hierarchies, and Transit Node Routing
        if func.__name__ in (
            "bidirectional_dijkstra",
            "contraction_hierarchy",
            "tnr",
        ):
            res = func(self, *args, **kwargs)
            if isinstance(res, dict) and "path" in res:
                res["path"] = self.__expand_path__(res["path"])
            return res

        # 3. One-sided algorithms with unreduced destination (or None)
        if destination_id is None or not is_reduced[destination_id]:
            res = func(self, *args, **kwargs)
            if isinstance(res, dict) and "path" in res:
                res["path"] = self.__expand_path__(res["path"])
            return res

        # 4. One-sided algorithm with reduced destination
        entry_nodes = self.reduced_inverse_graph[destination_id]
        if not entry_nodes:
            raise Exception(
                "Something went wrong, the origin and destination nodes are not connected."
            )

        best_length = float("inf")
        best_res = None
        best_entry = None

        for entry_u, entry_dist in entry_nodes.items():
            try:
                call_kwargs = dict(kwargs)
                call_args = list(args)
                if "destination_id" in call_kwargs:
                    call_kwargs["destination_id"] = entry_u
                elif len(call_args) > 1:
                    call_args[1] = entry_u

                res_u = func(self, *call_args, **call_kwargs)
                total_dist = res_u["length"] + entry_dist
                if total_dist < best_length:
                    best_length = total_dist
                    best_res = res_u
                    best_entry = entry_u
            except Exception:
                continue

        if best_res is None or best_length == float("inf"):
            raise Exception(
                "Something went wrong, the origin and destination nodes are not connected."
            )

        if "path" in best_res:
            expanded_path = self.__expand_path__(best_res["path"])
            tail = []
            if (
                self.reduced_inverse_graph_connections is not None
                and self.reduced_inverse_graph_connections[destination_id]
                is not None
            ):
                tail = self.reduced_inverse_graph_connections[
                    destination_id
                ].get(best_entry, [])
            full_path = expanded_path + tail + [destination_id]
            return {
                "path": full_path,
                "length": best_length,
            }
        return {
            "length": best_length,
        }

    def __dijkstra_on_graph__(
        self,
        graph: list[dict[int, float]],
        origin_id: int | set[int],
        destination_id: int,
    ) -> dict:
        """
        Function:

        - Internal Dijkstra solver on a specified graph dictionary list (used for same-chain routing on unreduced graph).
        """
        origin_ids = {origin_id} if isinstance(origin_id, int) else origin_id
        distance_matrix = [float("inf")] * len(graph)
        predecessor = [-1] * len(graph)
        open_leaves = []

        for oid in origin_ids:
            distance_matrix[oid] = 0
            heappush(open_leaves, (0, oid))

        while open_leaves:
            current_distance, current_id = heappop(open_leaves)
            if current_id == destination_id:
                break
            if current_distance == distance_matrix[current_id]:
                for (
                    connected_id,
                    connected_distance,
                ) in graph[current_id].items():
                    possible_distance = current_distance + connected_distance
                    if possible_distance < distance_matrix[connected_id]:
                        distance_matrix[connected_id] = possible_distance
                        predecessor[connected_id] = current_id
                        heappush(open_leaves, (possible_distance, connected_id))
        if current_id != destination_id:
            raise Exception(
                "Something went wrong, the origin and destination nodes are not connected."
            )

        return {
            "path": self.__reconstruct_path__(destination_id, predecessor),
            "length": distance_matrix[destination_id],
        }

    def __expand_path__(self, path: list[int]) -> list[int]:
        """
        Function:

        - Expand a path of reduced graph nodes to restore all intermediate pass-through nodes.

        Required Arguments:

        - `path`
            - Type: list[int]
            - What: The path in the reduced graph

        Returns:

        - The fully expanded path matching the original graph
        """
        fwd_conns = getattr(self, "reduced_graph_connections", None)
        inv_conns = getattr(self, "reduced_inverse_graph_connections", None)
        if fwd_conns is None and inv_conns is None:
            return path
        new_path = []
        append = new_path.append
        extend = new_path.extend
        for i in range(len(path) - 1):
            u = path[i]
            v = path[i + 1]
            append(u)
            expanded = False
            if fwd_conns is not None and fwd_conns[u] is not None:
                pass_throughs = fwd_conns[u].get(v)
                if pass_throughs:
                    extend(pass_throughs)
                    expanded = True
            if (
                not expanded
                and inv_conns is not None
                and inv_conns[v] is not None
            ):
                pass_throughs = inv_conns[v].get(u)
                if pass_throughs:
                    extend(pass_throughs)
        if path:
            append(path[-1])
        return new_path

    def is_same_chain(
        self,
        origin_id: int | set[int] | list[int],
        destination_id: int | None,
    ) -> bool:
        """
        Function:

        - Check whether any origin node and the destination node belong to the same reduced chain.

        Required Arguments:

        - `origin_id`
            - Type: int | set[int] | list[int]
            - What: The id(s) of the origin node(s)
        - `destination_id`
            - Type: int | None
            - What: The id of the destination node

        Returns:

        - True if both origin and destination are reduced nodes and share the same chain_id, False otherwise.
        """
        reduced_node_chain_ids = getattr(self, "reduced_node_chain_ids", None)
        if reduced_node_chain_ids is None or destination_id is None:
            return False
        if destination_id < 0 or destination_id >= len(reduced_node_chain_ids):
            return False
        dest_chain = reduced_node_chain_ids[destination_id]
        if dest_chain is None:
            return False
        if isinstance(origin_id, int):
            if 0 <= origin_id < len(reduced_node_chain_ids):
                return reduced_node_chain_ids[origin_id] == dest_chain
            return False
        for oid in origin_id:
            if (
                0 <= oid < len(reduced_node_chain_ids)
                and reduced_node_chain_ids[oid] == dest_chain
            ):
                return True
        return False
